core/ignore.js

```diff
@@ -13,152 +13,119 @@ export type IgnorePattern = {
+//
+// See https://git-scm.com/docs/gitignore/#_pattern_format
+class Ignore {
   /*::
   patterns: IgnorePattern[]
+  match: (string, boolean, IgnorePattern) => boolean
   */
 
   // Load patterns for detecting ignored files and folders
-  constructor (lines) {
-    this.patterns = []
-    for (let line of Array.from(lines)) {
-      if (line === '') { continue }          // Blank line
-      if (line[0] === '#') { continue }      // Comments
-      this.patterns.push(this.buildPattern(line))
-    }
-    autoBind(this)
-  }
-
-  // Parse a line and build the corresponding pattern
-  buildPattern (line) {
-    let folder = false
-    let negate = false
-    let noslash = line.indexOf('/') === -1
-    if (line.indexOf('**') !== -1) {   // Detect two asterisks
-      noslash = false
-    }
-    if (line[0] === '!') {               // Detect bang prefix
-      line = line.slice(1)
-      negate = true
-    }
-    if (line[0] === '/') {               // Detect leading slash
-      line = line.slice(1)
-    }
-    if (line[line.length - 1] === '/') {   // Detect trailing slash
-      line = line.slice(0, line.length - 1)
-      folder = true
-    }
-    line = line.replace(/^\\/, '')   // Remove leading escaping char
-    line = line.replace(/( |\t)*$/, '')  // Remove trailing spaces
-    // Ignore case for case insensitive file-systems
-    if (process.platform === 'darwin' || process.platform === 'win32') {
-      line = makeRe(line, {nocase: true})
-    }
-    let pattern = {
-      match: matcher(line, MicromatchOptions),
-      basename: noslash,   // The pattern can match only the basename
-      folder,    // The pattern will only match a folder
-      negate    // The pattern is negated
-    }
-    return pattern
+  constructor (lines /*: string[] */) {
+    this.patterns = Array.from(lines)
+      .filter(isNotBlankOrComment)
+      .map(buildPattern)
+    this.match = match
   }
 
   // Add some rules for things that should be always ignored (temporary
   // files, thumbnails db, trash, etc.)
   addDefaultRules () {
-    let morePatterns = (Array.from(DefaultRules).map((rule) => this.buildPattern(rule)))
+    // TODO: split on return char depending on the OS
+    const DefaultRules = fs
+      .readFileSync(resolve(__dirname, './config/.cozyignore'), 'utf8')
+      .split('\n')
+    let morePatterns = Array.from(DefaultRules).map(buildPattern)
     this.patterns = morePatterns.concat(this.patterns)
     return this
   }
 
-  // Return true if the doc matches the pattern
-  match (path, isFolder, pattern) {
-    if (pattern.basename) {
-      if (pattern.match(basename(path))) { return true }
-    }
-    if (isFolder || !pattern.folder) {
-      if (pattern.match(path)) { return true }
-    }
-    let parent = dirname(path)
-    if (parent === '.') { return false }
-    return this.match(parent, true, pattern)
-  }
-
   // Return true if the given file/folder path should be ignored
-  isIgnored (doc) {
+  isIgnored (doc /*: CozyDoc */) {
```

Comment:
Maybe use the `Metadata` type instead? (as defined in `core/metadata.js`)

Reply:
I looked for a type annotation that might be suitable for a `cozy-doc' but I couldn't find it. Thanks for letting me know. The naming doesn't seem to me very explicit.